### PR TITLE
Add consistency level to client and query options

### LINKED ISSUE #540: Ability to weaken consistency at client or request level
Add an option to set the client level consistency. 
Add an option to set consistency on query

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

Added client consistency level and made consistency level public on query.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Closing issues

Put closes #540

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -139,6 +140,12 @@ public Collection<RequestHandler> CustomHandlers
         /// <seealso cref="CosmosClientBuilder.WithConnectionModeGateway(int?)"/>
         public ConnectionMode ConnectionMode { get; set; }
 
+        /// <summary>
+        /// This can be used to weaken the database account consistency level for read operations.
+        /// If this is not set the database account consistency level will be used for all requests.
+        /// </summary>
+        public ConsistencyLevel? DesiredConsistencyLevel { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: >DesiredConsistencyLevel [](start = 33, length = 23)

Is name consistent in all places (Item, Query, client etc...)

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -171,7 +178,7 @@ public Collection<RequestHandler> CustomHandlers
         /// <summary>
         /// Gets the user json serializer with the CosmosJsonSerializerWrapper or the default
         /// </summary>
-        [JsonConverter(typeof(ClientOptionJsonConverter))]
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: >ClientOptionJsonConverter [](start = 30, length = 25)

Too late but why do we need it?

File: Microsoft.Azure.Cosmos/src/RequestOptions/RequestOptions.cs [hunk 3]
```diff
@@ -37,6 +37,19 @@ public class RequestOptions
...
+        /// Not every request supports consistency level. This allows each child to decide to expose it
+        /// but use the same base logic
+        /// </summary>
+        /// <value>
+        /// The consistency level required for the request.
+        /// </value>
+        /// <remarks>
+        /// ConsistencyLevel compatibility will validated and set by RequestInvokeHandler
+        /// </remarks>
+        internal virtual ConsistencyLevel? BaseConsistencyLevelHelper { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/RequestOptions/RequestOptions.cs]: >BaseConsistencyLevelHelper [](start = 43, length = 26)

Is it helper?

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 4]
```diff
@@ -207,5 +182,43 @@ private void FillMultiMasterContext(RequestMessage request)
...
+            }
+
+            if (consistencyLevel.HasValue)
+            {
+                if (!this.AccountConsistencyLevel.HasValue)
+                {
+                    this.AccountConsistencyLevel = await this.client.GetAccountConsistencyLevelAsync();
+                }
+                
+                if (ValidationHelpers.ValidateConsistencyLevel(this.AccountConsistencyLevel.Value, consistencyLevel.Value))
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs]: >ValidateConsistencyLevel [](start = 38, length = 24)

IsValidOverwrite

File: Microsoft.Azure.Cosmos/src/RequestOptions/RequestOptions.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/RequestOptions/RequestOptions.cs]: I guess not.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: We do a serialization of the settings, and the standard serializer will throw on this.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Updated it to match item, query, client. It's now just ConsistencyLevel

**Comment by kirankumarkolli** [comment]
ChangeLog and usage sample update please.

**Pull request merged**